ffi/newpassmanagers.cpp

```diff
@@ -0,0 +1,317 @@
+#include "core.h"
+#include "llvm-c/TargetMachine.h"
+#include "llvm/Analysis/AliasAnalysisEvaluator.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Transforms/InstCombine/InstCombine.h"
+#include "llvm/Transforms/Scalar/JumpThreading.h"
+#include "llvm/Transforms/Scalar/LoopRotation.h"
+#include "llvm/Transforms/Scalar/LoopUnrollPass.h"
+#include "llvm/Transforms/Scalar/SimplifyCFG.h"
+
+using namespace llvm;
+
+/*
+ * Exposed API
+ */
+
+namespace llvm {
+
+struct OpaqueModulePassManager;
+typedef OpaqueModulePassManager *LLVMModulePassManagerRef;
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(ModulePassManager, LLVMModulePassManagerRef)
+
+struct OpaqueFunctionPassManager;
+typedef OpaqueFunctionPassManager *LLVMFunctionPassManagerRef;
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(FunctionPassManager,
+                                   LLVMFunctionPassManagerRef)
+
+struct OpaquePassBuilder;
+typedef OpaquePassBuilder *LLVMPassBuilderRef;
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(PassBuilder, LLVMPassBuilderRef)
+
+struct OpaquePipelineTuningOptions;
+typedef OpaquePipelineTuningOptions *LLVMPipelineTuningOptionsRef;
+DEFINE_SIMPLE_CONVERSION_FUNCTIONS(PipelineTuningOptions,
+                                   LLVMPipelineTuningOptionsRef)
+
+static TargetMachine *unwrap(LLVMTargetMachineRef P) {
+    return reinterpret_cast<TargetMachine *>(P);
+}
+
+} // namespace llvm
+
+extern "C" {
+
+// MPM
+
+API_EXPORT(LLVMModulePassManagerRef)
+LLVMPY_CreateNewModulePassManager() {
+    return llvm::wrap(new PassManager<Module>());
+}
+
+API_EXPORT(void)
+LLVMPY_NMPRun_module(LLVMModulePassManagerRef MPMRef, LLVMPassBuilderRef PBRef,
+                     LLVMModuleRef mod) {
+
+    ModulePassManager *MPM = llvm::unwrap(MPMRef);
+    PassBuilder *PB = llvm::unwrap(PBRef);
+    Module *M = llvm::unwrap(mod);
+
+    LoopAnalysisManager LAM;
+    FunctionAnalysisManager FAM;
+    CGSCCAnalysisManager CGAM;
+    ModuleAnalysisManager MAM;
+    PB->registerLoopAnalyses(LAM);
+    PB->registerFunctionAnalyses(FAM);
+    PB->registerCGSCCAnalyses(CGAM);
+    PB->registerModuleAnalyses(MAM);
+    PB->crossRegisterProxies(LAM, FAM, CGAM, MAM);
```

Comment:
Can you run a module pass manager multiple times? If so, what are the semantics WRT these analysis managers? They appear to go out of scope at the end of this function, but does the `PassBuilder` consume them? Does it expect them to outlive it? What happens when analysis managers are registered a second time?

Reply:
>Can you run a module pass manager multiple times? If so, what are the semantics WRT these analysis managers?

Yes we can, we will invoke this function again. Re-register all the analyses passes and then run the pipeline.
> They appear to go out of scope at the end of this function, but does the PassBuilder consume them? Does it expect them to outlive it?

They indeed go out of scope and PassBuilder doesn't consume/store them. The reason why we are creating all these analyses managers is various analyses passes can be registered only under their own analyses managers. For eg there will be some loop analyses registered under LoopAnalysesManager, but while optimizing a module we need the ability to run these loop analyses on loops inside the module. So we register analyses passes in under their respective managers and cross register them with each other to solve the problem I described above.

>What happens when analysis managers are registered a second time?

You are free to create as many managers as you want and they are independent of each other. All you need is populated + cross registered corresponding analysis manager whenever invoking `run` on the pass manager (FAM for FPM, MAM for MPM).